xds/src/main/java/io/grpc/xds/XdsServerWrapper.java

```diff
@@ -0,0 +1,439 @@
+        internalStart();
+      }
+    });
+    syncContext.drain();
+    IOException exception;
+    try {
+      exception = initialStartFuture.get();
+    } catch (InterruptedException | ExecutionException e) {
+      throw new RuntimeException(e);
+    }
+    if (exception != null) {
+      throw exception;
+    }
+    return this;
+  }
+
+  private void internalStart() {
+    try {
+      xdsClientPool = xdsClientPoolFactory.getOrCreate();
+    } catch (Exception e) {
+      listener.onNotServing(
+          Status.UNAVAILABLE.withDescription(
+              "Failed to initialize xDS").withCause(e).asException());
+      initialStartFuture.set(null);
+      return;
+    }
+    xdsClient = xdsClientPool.getObject();
+    // TODO(chengyuanzhang): add an API on XdsClient indicating if it is using v3, don't get
+    //  from bootstrap.
+    boolean useProtocolV3 = xdsClient.getBootstrapInfo().getServers().get(0).isUseProtocolV3();
+    String listenerTemplate = xdsClient.getBootstrapInfo().getServerListenerResourceNameTemplate();
+    if (!useProtocolV3 || listenerTemplate == null) {
+      listener.onNotServing(
+          Status.UNAVAILABLE.withDescription(
+              "Can only support xDS v3 with listener resource name template").asException());
+      initialStartFuture.set(null);
+      xdsClient = xdsClientPool.returnObject(xdsClient);
+      return;
+    }
+    discoveryState = new DiscoveryState(listenerTemplate.replaceAll("%s", listenerAddress));
+  }
+
+  @Override
+  public Server shutdown() {
+    if (!shutdown.compareAndSet(false, true)) {
+      return this;
+    }
+    syncContext.execute(new Runnable() {
+      @Override
+      public void run() {
+        internalShutdown();
+        delegate.shutdown();
+      }
+    });
+    syncContext.drain();
+    return this;
+  }
+
+  @Override
+  public Server shutdownNow() {
+    if (!shutdown.compareAndSet(false, true)) {
+      return this;
+    }
+    syncContext.execute(new Runnable() {
+      @Override
+      public void run() {
+        internalShutdown();
+        delegate.shutdownNow();
+      }
+    });
+    syncContext.drain();
+    return this;
+  }
+
+  // Must run in SynchronizationContext
+  private void internalShutdown() {
+    logger.log(Level.FINER, "Shutting down XdsServerWrapper");
+    if (discoveryState != null) {
+      discoveryState.shutdown();
+    }
+    if (xdsClient != null) {
+      xdsClient = xdsClientPool.returnObject(xdsClient);
+    }
+    if (timeService != null) {
+      SharedResourceHolder.release(GrpcUtil.TIMER_SERVICE, timeService);
+    }
+    if (restartTimer != null) {
+      restartTimer.cancel();
+    }
+    isServing = false;
+    internalTerminationLatch.countDown();
+  }
+
+  @Override
+  public boolean isShutdown() {
+    return shutdown.get();
+  }
+
+  @Override
+  public boolean isTerminated() {
+    if (internalTerminationLatch.getCount() != 0) {
+      return false;
+    }
+    return delegate.isTerminated();
```

Comment:
With the current implementation (order) it is enough to check `delegate.isTerminated()`, without the lines 209-211, isn't it?

But if we change the order in `shutdown` as per my suggestion, do we still need to check `internalTermination` status?